### PR TITLE
Rebase on goodtables.io API

### LINKED ISSUE #14: Rebase on goodtables.io
# Overview

We have two options of updating this library:
- use https://github.com/frictionlessdata/goodtables-server-py as inspection provider
- add file uploading support to `goodtables.io` and use this service as inspection provider

### PR DESCRIPTION
**Pull request by roll**
- fixes #14

---

Source tree - https://github.com/frictionlessdata/goodtables-js/tree/goodtables-io-rebase

---

For now it's mostly POC/MVP. After `try.goodtables.js` initial release I'm going to improve readme, tests etc as part of #15 

---

API decision:
- provide only `goodtables.validate()` method which returns `report` 
  - we will add the same high-level wrapper function for `goodtables-py`
  - initial decision to introduce `inspect` concept was because our other libs `validate` functions raise on invalid input. For now I see that it was probably a mistake. E.g. in goodtables.io work we always had those confusion between inspection and validation. And talking with people and community work we still name it validation. So now I think that just drop one unnecessary term (inspect) will be a good idea.
  - we could have this behavior across libs - https://github.com/frictionlessdata/implementations/commit/a87f2c988f05c4cdcd72925dd459345e420c15f3 (we of course could skip this step but in this case having `validate` with different behavior could be a little bit confusing)
- provide `goodtables.inspect()` returning report and `goodtables.validate()` raising error (second function is optional)
- provide only `goodtables.validate()` function (but TBH I don't think that this try/catch or try/except interface is really nice to user on any Python or JavaScript. With try/except as a user you just write additional code to get report/errors):
```js
try {
  report = goodtables.validate()
} catch (error) {
  report = error.report
}
```

### CODE REVIEW

**Comment by roll** [comment]
@pwalsh 
@amercader 
@akariv 
please take a look

**Comment by pwalsh** [comment]
The API discussion is really about goodtables API in general, and not this library, so my comments are only about that:

1. goodtables errors are a fundamentally different thing to exceptions. I feel like there is some confusion between the two in what I read above, and in https://github.com/frictionlessdata/implementations/commit/a87f2c988f05c4cdcd72925dd459345e420c15f3 . Some goodtables errors are generated from exceptions (e.g: io errors), but that doesn't mean they are the same thing.
2. Based on (1), I don't see the conflict with other libraries raising on exceptions, and goodtables returning a report of errors, whether the method happens to be named the same or not.
3. Also based on (1), on your comment on try/except not being nice in Python: it is just a very subjective statement that it is at odds with Pythonic approaches to exceptions. I'm trying to understand if there is confusion here around errors (for humans in goodtables) and exceptions (as a programming construct), or just that you don't like idioms around exception handling in Python.

Overall, I don't mind what you decide, but I think it is important not to overthink this :).

**Comment by pwalsh** [comment]
Comments for the PR here, API discussion aside - looks good!

**Comment by pwalsh** [comment]
Also @roll on `inspect` vs. `validate`, I don't really care either way anymore, but I don't think you made a "mistake" here - the idea of the `Inspector` class is solid and conceptually useful, for example, and whether the method called to get a report is `inspect` or `validate` is AOK with me.

File: src/validate.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+const pickBy = require('lodash/pickBy')
+const mapKeys = require('lodash/mapKeys')
+const isString = require('lodash/isString')
+const snakeCase = require('lodash/snakeCase')
+const config = require('./config')
+
+
+// Module API
+
+module.exports.validate = async (source, options) => {
```
Comment (amercader) [inline src/validate.js:12]: What are you supposed to pass as `source`? The file URL?

**Comment by amercader** [comment]
@roll The PR looks good, obviously needs more documentation and examples as you said.

I'm not sure I understand the difference between Inspect and Validate, probably because I'm missing some context. Is it just a matter of supporting `validate(raise=True/False)`?

File: src/validate.js (hunk 1 above)
Comment (roll) [reply src/validate.js:12]: Adria it's the same as in `goodtables-py` and our API. Url or list of urls or list of datapackages (based on preset) etc. So it definitely requires perfect documentation :smiley:

**Comment by roll** [comment]
Thanks everyone. So let's have it as it is for now (as `0.4.0`) to start `try.goodtables.io` and thanks to `npm` behavior we could re-review API (e.g. with @amercader idea) a little bit later (when will be releasing this lib and `goodtables-py`) because even breaking changes is non-damaging for Node (e.g. on `0.4` -> `0.5`).

**Pull request merged**